node/rpc/src/lib.rs

```diff
@@ -28,19 +52,16 @@ where
 {
     pub state_manager: Arc<StateManager<DB>>,
     pub keystore: Arc<RwLock<KS>>,
+    pub heaviest_tipset: Arc<RwLock<Option<Arc<Tipset>>>>,
+    pub subscriber: Subscriber<HeadChange>,
```

Comment:
If anything, this should be an `Arc<Publisher<..>>` where if you need to get something to subscribe to it, you use that to create a new subscriber

Reply:
The reason why this is a subscribe is that it is a one time usage just for the purpose of the sub_head_changes function. We are only ever using this to poll from the subscribe channel to iterate our HeadChangeJson.

Having a publisher return subscribers will mean that publisher will have to be behind a Arc<Lock which also carries overhead

Reply:
but this RpcState is holding this subscriber that never gets polled, as far as I can see (can you point to where it's being handled?).

The only reason you would hold the publisher is so that you'd be able to subscribe without holding a dead subscriber. I do see the overhead concern with this, especially since the subscribe would require a write lock for some reason, but as of now that would be better than the node stalling after this subscriber buffer fills